### ISSUE TITLE
[v5] deletion of root key in reducer isn't persisted

### ISSUE THREAD

**Issue by Zycon42** [rt2zz/redux-persist#492]
version: 5.0.0-rc.6

In library such as redux-form they use whole state as dictionary and each form is root key in state such as
```
{
  form1: <form state>,
  form2: <form state>
}
```

when form is destroyed they delete corresponding root key from state. Unfortunately this isn't detected by redux-persist. Key is deleted from state but after reload it's rehydrated.

I traced problem to the persistoid update function which doesn't detect keys that were deleted from state and it doesn't schedule them for deletion inside storage.

**Comment by rt2zz** [comment]
hm, this is a difficult issue, because there are some cases where you *want* the removed reducer state to stay in storage, e.g. if you are adding / removing reducers dynamically (e.g. code splitting).

Is it possible to have redux-form nested one level deeper, i.e. 
```js
state = {
  form: { 
    form1: {...},
  },
}
```

If it needs to be top level state, then I think we need to think about an api change or addition that will make deleting state possible without making it impossible to have dynamically loaded persisted reducers.

**Comment by Zycon42** [comment]
I'm not sure if I made it clear but I don't have redux-form reducer as top level reducer it's mounted in `state.form` already where state is redux root. Changing default location of redux-form from `state.form` is supported but that makes usage very cumbersome. In v4 diffing is made for the whole state tree so it doesn't intervene with redux-form.

Would it be possible to add some flag to persistReducer config that will turn this behaviour off to allow deleting reducer keys?

**Comment by rt2zz** [comment]
I see, and is your persistReducer wrapped around the root reducer or the form reducer?

Assuming persistReducer is around the root reducer, it should be storing the updated state *without* the removed keys. i.e. it should work like you expected it to work. If that is not the case then I think we will need to work up a test case and figure out what is going on.

In the case that you want to wrap just the form reducer with persistReducer, then yes it sounds like we need to think about how to add this feature without disrupting support for code splitting / lazy loaded reducers.

Ill start thinking on this a bit.

**Comment by Zycon42** [comment]
Yes I have formReducer wrapped with persistReducer. I have it like this to use different storages for different reducers (auth keys go to keychain, rest to AsyncStorage).

Thank you very much for your assistance.

**Comment by rt2zz** [comment]
ok I think I was mistaken earlier, the code split case is handled via second persistedReducer, so we should in fact be removing removed keys when they are removed from state. We will need to double check that no known use cases are impacted by this change, but assuming not I think we should do this change on v5.

In the short term, you might consider pulling your form persistReducer up to the top level with `whitelist: ['form']` in config. Note however: if you do go this route short term, there will be no easy state migration to get back to the nested persistReducer form.

**Comment by rt2zz** [comment]
@Zycon42 this may have been resolved in 5.1.0

**Comment by chawax** [comment]
I had exactly the same problem as @Zycon42 with 5.9.1 so it looks like it has not been solved.

I solved it by purging manually the persisted state :

```
export const formPersistConfig = {
  key: 'form',
  storage: AsyncStorage,
  whitelist: ['subscriptionForm']
}
```

```
import { purgeStoredState } from 'redux-persist'
import { formPersistConfig } from '../../../ReduxPersistConfig'
purgeStoredState(formPersistConfig)
```

It works because I have only one form in my whitelist, but not sure how it will work if I have more forms.

**Comment by Dante-101** [comment]
@chawax I had the same issue. I edited the code and it works fine for my case now. See if this helps you too - https://github.com/rt2zz/redux-persist/issues/811#issuecomment-384983586
